### PR TITLE
Fix issue where retrieving a twin array from registry manager returned null

### PR DESCRIPTION
**Pull request by abhipsaMisra**
On retrieving a twin property by index/ key, our sdk will try to return it as a `JObject` within a `TwinCollection`. For twin arrays, since they are represented as `JArray`, they cannot be converted to a `JObject` (`JArray` is not inherited from `JObject`), and are thus returned as `null`. 
This PR adds a new `TwinCollectionArray` type, one that is inherited from a `JArray`, and returns that instead`.

Note: this only affects the registry manager retuned twin. The device-client returned twin is always returned as the base `JToken` type: https://github.com/Azure/azure-iot-sdk-csharp/blob/abmisr/twinArrayNull/shared/src/TwinCollection.cs#L255

### CODE REVIEW

File: shared/src/TwinCollectionArray.cs [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+            : base(jArray)
+        {
+            _metadata = metadata;
+        }
+
+        /// <summary>
+        /// Gets the value for the given property name
+        /// </summary>
+        /// <param name="propertyName">Property Name to lookup</param>
+        /// <returns>Value if present</returns>
```
Comment (azabbasi) [inline shared/src/TwinCollectionArray.cs:30]: nit: `Property value if present`

File: shared/src/TwinCollectionArray.cs [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+                {
+                    return GetLastUpdated();
+                }
+
+                if (propertyName == LastUpdatedVersionName)
+                {
+                    return GetLastUpdatedVersion();
+                }
+
+                throw new ArgumentException($"'Newtonsoft.Json.Linq.JArray' does not contain a definition for '{propertyName}'.");
```
Comment (azabbasi) [inline shared/src/TwinCollectionArray.cs:50]: is it really the Newtonsoft object that doesn't have a definition for the property? or the twin object?

File: shared/src/TwinCollectionArray.cs (hunk 1 above)
Comment (drwill-ms) [reply shared/src/TwinCollectionArray.cs:30]: And add a comma before "if present"

File: shared/src/TwinCollectionArray.cs [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+        private const string MetadataName = "$metadata";
+        private const string LastUpdatedName = "$lastUpdated";
+        private const string LastUpdatedVersionName = "$lastUpdatedVersion";
+
+        private readonly JObject _metadata;
+
+        internal TwinCollectionArray(JArray jArray, JObject metadata)
+            : base(jArray)
+        {
+            _metadata = metadata;
```
Comment (drwill-ms) [inline shared/src/TwinCollectionArray.cs:23]: null check?

File: shared/src/TwinCollectionArray.cs [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+using Newtonsoft.Json.Linq;
+
+namespace Microsoft.Azure.Devices.Shared
+{
+    /// <summary>
+    /// Represents a property array in <see cref="TwinCollection"/>
+    /// </summary>
+    public class TwinCollectionArray : JArray
+    {
+        private const string MetadataName = "$metadata";
```
Comment (drwill-ms) [inline shared/src/TwinCollectionArray.cs:14]: Can you find a way to not have to duplicate these constants across files?

File: shared/src/TwinCollectionValue.cs [hunk 5]
```diff
@@ -17,9 +17,9 @@ namespace Microsoft.Azure.Devices.Shared
         Justification = "Uses default JValue comparison, equality and hashing implementations.")]
     public class TwinCollectionValue : JValue
     {
-        const string metadataName = "$metadata";
-        const string lastUpdatedName = "$lastUpdated";
-        const string lastUpdatedVersionName = "$lastUpdatedVersion";
+        private const string metadataName = "$metadata";
```
Comment (drwill-ms) [inline shared/src/TwinCollectionValue.cs:20]: Field casing is wrong for constants

**Review by drwill-ms — CHANGES_REQUESTED** [review]
Waiting on responses to code comments.

File: shared/src/TwinCollectionArray.cs (hunk 1 above)
Comment (abhipsaMisra) [reply shared/src/TwinCollectionArray.cs:30]: will update.

File: shared/src/TwinCollectionArray.cs (hunk 2 above)
Comment (abhipsaMisra) [reply shared/src/TwinCollectionArray.cs:50]: yeah, it should be ~TwinCollectionArray` actually, i'll update it

File: shared/src/TwinCollectionArray.cs (hunk 3 above)
Comment (abhipsaMisra) [reply shared/src/TwinCollectionArray.cs:23]: metadata is a JObject, which will default to null anyway. Is there an advantage to adding an explicit null check here?

File: shared/src/TwinCollectionArray.cs (hunk 4 above)
Comment (abhipsaMisra) [reply shared/src/TwinCollectionArray.cs:14]: sure, will do

File: shared/src/TwinCollectionArray.cs [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+
+                throw new ArgumentException($"{nameof(TwinCollectionArray)} does not contain a definition for '{propertyName}'.");
+            }
+        }
+
+        /// <summary>
+        /// Gets the Metadata for this property
+        /// </summary>
+        /// <returns>Metadata instance representing the metadata for this property</returns>
+        public Metadata GetMetadata()
```
Comment (abhipsaMisra) [inline shared/src/TwinCollectionArray.cs:54]: this, and below: they are common for both TwinCollectionArray and TwinCollectionValue; but with multiple inheritance not supported in c#, I've kept them repeated. I could define an interface with implementation, and use that, but it seemed too much of work for this.
What do you think?

File: shared/src/TwinCollectionArray.cs (hunk 3 above)
Comment (drwill-ms) [reply shared/src/TwinCollectionArray.cs:19]: There is no guarding against NRE later in this file.

So either it can be null, in which case we we should conditionally derereference it, or we should check for null and throw.

File: shared/src/TwinCollectionArray.cs [hunk 7]
```diff
@@ -0,0 +1,77 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using Newtonsoft.Json.Linq;
+
+namespace Microsoft.Azure.Devices.Shared
+{
+    /// <summary>
+    /// Represents a property array in <see cref="TwinCollection"/>
```
Comment (drwill-ms) [inline shared/src/TwinCollectionArray.cs:10]: in a

File: shared/src/TwinCollectionArray.cs (hunk 6 above)
Comment (drwill-ms) [reply shared/src/TwinCollectionArray.cs:54]: I don't feel too much code is duplicated. If you want to pursue a common base class, that'd be fine, but I don't expect it.

Comment (abhipsaMisra) [reply shared/src/TwinCollectionArray.cs:54]: I am going to leave it as-is. It isn't a whole lot of repeated code and I don't expect it to grow in the future either.

File: shared/src/TwinCollectionArray.cs (hunk 3 above)
Comment (abhipsaMisra) [reply shared/src/TwinCollectionArray.cs:19]: I see, I can throw an ArgumentNullException during initialization, sure. The use-case does require the metadata to be available (and this is an internal constructor), so makes sense.

**Review by azabbasi — APPROVED** [review]

File: shared/src/TwinCollectionArray.cs [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+            return new Metadata(GetLastUpdated(), GetLastUpdatedVersion());
+        }
+
+        /// <summary>
+        /// Gets the LastUpdated time for this property
+        /// </summary>
+        /// <returns>DateTime instance representing the LastUpdated time for this property</returns>
+        public DateTime GetLastUpdated()
+        {
+            return (DateTime)_metadata[TwinCollection.LastUpdatedName];
```
Comment (drwill-ms) [inline shared/src/TwinCollectionArray.cs:65]: I think the return value needs to change to `DateTime?` and the return statement to `return (DateTime?)_metadata?[TwinCollection.LastUpdatedName];`

File: shared/src/TwinCollectionArray.cs [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+        }
+
+        /// <summary>
+        /// Gets the LastUpdatedVersion for this property
+        /// </summary>
+        /// <returns>LastUpdatdVersion if present, null otherwise</returns>
+        public long? GetLastUpdatedVersion()
+        {
+            return (long?)_metadata[TwinCollection.LastUpdatedVersionName];
```
Comment (drwill-ms) [inline shared/src/TwinCollectionArray.cs:74]: should be `return (long?)_metadata?[TwinCollection.LastUpdatedVersionName];`

File: shared/src/TwinCollectionArray.cs (hunk 8 above)
Comment (drwill-ms) [reply shared/src/TwinCollectionArray.cs:65]: Would you add a couple of unit tests to validate that this methods down throw when initialized with null?

**Review by drwill-ms — APPROVED** [review]

**Review by barustum — APPROVED** [review]

**Comment by ThomasPe** [comment]
any ETA when this fix will be released to nuget?

**Comment by abhipsaMisra** [comment]
We are in the process of finalizing our next release. We should have it out by end of next week.
Does this issue currently block you, @ThomasPe ?

**Comment by ThomasPe** [comment]
Indeed it does, the earlier the better. Appreciate your efforts 👍

**Comment by abhipsaMisra** [comment]
Sure thing, I'll keep you updated!
